### PR TITLE
Add BmiClientApptainer

### LINKED ISSUE #116: Apptainer client
As apptainer and singularity are diverging it would be nice that grpc4bmi has first class support for apptainer. Since #111 you can use apptainer in BmiClientSingularity as apptainer creates a singularity symlink, but this make it confusing to tell whether singularity or apptainer is used.

What I would like to do is

```python
from grpc4bmi.bmi_client_apptainer import BmiClientApptainer
client = BmiClientApptainer(image='docker://ewatercycle/marrmot-grpc4bmi:latest')
```

instead of 

```python
from grpc4bmi.bmi_client_singularity import BmiClientSingularity
client = BmiClientSingularity(image='docker://ewatercycle/marrmot-grpc4bmi:latest')
```

### PR DESCRIPTION
**Pull request by sverhoeven**
Adds `grpc4bmi.bmi_client_apptainer.BmiClientApptainer` class. 

Fixes #116

Requires a https://apptainer.org installation to run tests.

After this PR is merged a new version on the 0.2.x branch should be published to GitHub/PyPi/Zenodo.

Due to Singularity and Apptainer being so similar, the code duplication has increased, this will disappear once we deprecate singularity.

### CODE REVIEW

File: grpc4bmi/bmi_client_apptainer.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+            The temporary file gets removed when this object is deleted.
+
+    **Example 1: Config file already inside image**
+
+    MARRMoT has an `example config file <https://github.com/wknoben/MARRMoT/blob/master/BMI/Config/BMI_testcase_m01_BuffaloRiver_TN_USA.mat>`_ inside its Docker image.
+
+    .. code-block:: python
+
+        from grpc4bmi.bmi_client_apptainer import BmiClientApptainer
+        client = BmiClientApptainer(image='docker://ewatercycle/marrmot-grpc4bmi:latest')
```
Comment (BSchilperoort) [inline grpc4bmi/bmi_client_apptainer.py:102]: This line won't run. work_dir is a required argument;
`TypeError: __init__() missing 1 required positional argument: 'work_dir'`

Comment (BSchilperoort) [reply grpc4bmi/bmi_client_apptainer.py:102]: The next 3 lines run fine if a work_dir is given.

File: grpc4bmi/bmi_client_docker.py [hunk 2]
```diff
@@ -23,7 +23,7 @@ class BmiClientDocker(BmiClient):
         host (str): Host on which the image port is published on a random port
         input_dirs (Iterable[str]): Directories for input files of model.
 
-            All of directories will be mounted read-only inside Singularity container on same path as outside container.
+            All of directories will be mounted read-only inside Docker container on same path as outside container.
```
Comment (BSchilperoort) [inline grpc4bmi/bmi_client_docker.py:26]: Well spotted! :)

**Review by BSchilperoort — APPROVED** [review]
Hi Stefan, very nice work. I was able to run the (`test_apptainer`) tests locally (on WSL2). The design overview graph looks a lot better 👍 

I did notice a small mistake in a docstring example though (see the specific comment), but I'll approve the PR already.

Additionally I wondered about the docs generation, it seems that the readthedocs CI is not linked to this repository? It makes going through the API docstring a bit harder, and perhaps some of the docs changes break rtd.

**Comment by sverhoeven** [comment]
Thanks for reviewing. 

I fixed the docstring example, it was also present in the BmiClientSingularity docstring.

The https://grpc4bmi.readthedocs.io/en/0.2.x/ gets updated each 0.2.x merge. I updated the docs badge to point to that.

**Comment by sverhoeven** [comment]
v0.2.16 has been released.

**Pull request merged**